### PR TITLE
Rest Client Reactive: Add reactive flavor for ClientHeadersFactory

### LINKED ISSUE #20001: Add a reactive ClientHeadersFactory flavor
### Description

Currently, org.eclipse.microprofile.rest.client.ext.ClientHeadersFactory is used to intercept calls and add entries in the client request header.

A very common scenario is to integrate with an api that is secured by a token, that is generated by calling a specific endpoint. The current approach does not support this scenario, because it expects a sync response.

We need to support this use case.

### Implementation ideas

A good way to implement this would be to create something similar to org.jboss.resteasy.reactive.server.ServerRequestFilter.

That is an annotation that supports many kinds of responses. We could create a ClientRequestFilter annotation that would support the client request context in the method parameters and allow returning a Uni, so the request could be updated in an async way.

### PR DESCRIPTION
**Pull request by Sgitario**
Added reactive flavor of `ClientHeadersFactory` that allows doing blocking operations. For example:

```java
package org.acme.rest.client;

import org.eclipse.microprofile.rest.client.ext.ClientHeadersFactory;

import javax.enterprise.context.ApplicationScoped;
import javax.ws.rs.core.MultivaluedHashMap;
import javax.ws.rs.core.MultivaluedMap;
import java.util.UUID;

@ApplicationScoped
public class GetTokenReactiveClientHeadersFactory extends ReactiveClientHeadersFactory {

    @Inject
    Service service;

    @Override
    public Uni<MultivaluedMap<String, String>> getHeaders(MultivaluedMap<String, String> incomingHeaders) {
        return Uni.createFrom().item(() -> {
            MultivaluedHashMap<String, String> newHeaders = new MultivaluedHashMap<>();
            // perform blocking call
            newHeaders.add(HEADER_NAME, service.getToken());
            return newHeaders;
        });
    }
}
```

Resolves https://github.com/quarkusio/quarkus/issues/20001

### CODE REVIEW

**Comment by Sgitario** [comment]
@michalszynkiewicz this PR proposes to create a version of ClientHeadersFactory for reactive by implementing the `ClientHeadersFactory` interface. 

The benefit is that we can reuse the same annotation (`@RegisterClientHeaders`) and features that were already in place at `MicroProfileRestClientRequestFilter.java`.
However, the new reactive flavor of ClientHeadersFactory, `ReactiveClientHeadersFactory` will NOT be an interface, but an abstract class, and also I had to implement a fake `update` method (see https://github.com/quarkusio/quarkus/pull/21807/files#diff-588a0da8278ac069a69d2db5ef55e2c8e165b1995e667b24550731e53310afd7R16). I'm throwing an exception to make a strong notice that this method should never be called plus I also disallow users to overwrite it.

In summary, by extending `ClientHeadersFactory`, we reuse everything that was in place at `MicroProfileRestClientRequestFilter.java`, but we are forced to implement the `update` method in  `ReactiveClientHeadersFactory`. 

If you're not happy with, the alternative is to:
- Create a new annotation `@RegisterReactiveClientHeaders`, with a default implementation `DefaultReactiveClientHeadersFactoryImpl` that does nothing, 
- Add the reactive filter and copy/somehow reuse everything that was in `MicroProfileRestClientRequestFilter.java`.
- Update the deployments to work with the new annotation which will imply to duplicate a few things.

Let me know your thoughts about it.

File: extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/ReactiveClientHeadersFactory.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * Reactive ClientHeadersFactory flavor for Quarkus rest-client reactive extension.
+ */
+public abstract class ReactiveClientHeadersFactory implements ClientHeadersFactory {
+    public abstract Uni<MultivaluedMap<String, String>> getHeaders(MultivaluedMap<String, String> incomingHeaders);
+
+    @Override
+    public final MultivaluedMap<String, String> update(MultivaluedMap<String, String> incomingHeaders,
+            MultivaluedMap<String, String> clientOutgoingHeaders) {
+        throw new RuntimeException(
+                "Can't call `update` method in a Reactive context. Use `getHeaders` or implements ClientHeadersFactory.");
```
Comment (michalszynkiewicz) [inline extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/ReactiveClientHeadersFactory.java:19]:
```suggestion
                "Can't call `update` method in a Reactive context. Use `getHeaders` or implement ClientHeadersFactory.");
```

File: extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/ReactiveClientHeadersFactory.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import javax.ws.rs.core.MultivaluedMap;
+
+import org.eclipse.microprofile.rest.client.ext.ClientHeadersFactory;
+
+import io.smallrye.mutiny.Uni;
+
+/**
+ * Reactive ClientHeadersFactory flavor for Quarkus rest-client reactive extension.
+ */
+public abstract class ReactiveClientHeadersFactory implements ClientHeadersFactory {
```
Comment (michalszynkiewicz) [inline extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/ReactiveClientHeadersFactory.java:12]: do we need to extend `ClientHeadersFactory`?

File: extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/MicroProfileRestClientRequestFilter.java [hunk 3]
```diff
@@ -63,7 +65,21 @@ public void filter(ClientRequestContext requestContext) {
         }
 
         if (clientHeadersFactory != null) {
-            incomingHeaders = clientHeadersFactory.update(incomingHeaders, headers);
+            if (clientHeadersFactory instanceof ReactiveClientHeadersFactory
+                    && requestContext instanceof ResteasyReactiveClientRequestContext) {
```
Comment (michalszynkiewicz) [inline extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/MicroProfileRestClientRequestFilter.java:69]: this filter should always work with `RestEasyReactiveClientRequestContext`, no? I don't think we need to check it with `instanceof`

File: extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/ReactiveClientHeadersFactory.java (hunk 2 above)
Comment (Sgitario) [reply extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/ReactiveClientHeadersFactory.java:12]: Yes, to be used with `@RegisterClientHeaders`.

File: extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/MicroProfileRestClientRequestFilter.java (hunk 3 above)
Comment (Sgitario) [reply extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/MicroProfileRestClientRequestFilter.java:69]: I think so... but I thought we could be extra safe here. However, will remove the instanceof.

**Review by michalszynkiewicz — APPROVED** [review]
thanks @Sgitario !

File: docs/src/main/asciidoc/rest-client-reactive.adoc [hunk 4]
```diff
@@ -591,6 +591,37 @@ To specify a value for `${header.value}`, simply put the following in your `appl
...
+
+    @Inject
+    Service service;
+
+    @Override
+    public Uni<MultivaluedMap<String, String>> getHeaders(MultivaluedMap<String, String> incomingHeaders) {
+        return Uni.createFrom().item(() -> {
+            MultivaluedHashMap<String, String> newHeaders = new MultivaluedHashMap<>();
+            // perform blocking call
+            newHeaders.add(HEADER_NAME, service.getToken());
```
Comment (SimonScholz) [inline docs/src/main/asciidoc/rest-client-reactive.adoc:618]: Hi,
I am sorry that I kind of hijack this already merged PR, but IMHO the `ReactiveClientHeadersFactory` does not work as expected.
Why has the `service.getToken()` to be a blocking call?
In my case I wanted to call the auth server where I obtain a token also in a reactive manner by using an reactive rest client.
Therefore my code would look like this:
```
    override fun getHeaders(incomingHeaders: MultivaluedMap<String, String>?): Uni<MultivaluedMap<String, String>> {
        // Utilizing the reactive mutiny api :-), which unfortunately does not work....
        return clientTokenFetcher.getClientToken().onItem().transform {
            MultivaluedHashMap<String, String>().apply {
                add(HttpHeaders.CACHE_CONTROL, "no-cache")
                add(HttpHeaders.CONTENT_TYPE, "text/plain")
                add(HttpHeaders.AUTHORIZATION, it.authHeader)
            }
        }
    }
``` 

Unfortunately this does not work, while this is working:

```
    override fun getHeaders(incomingHeaders: MultivaluedMap<String, String>?): Uni<MultivaluedMap<String, String>> {
        return Uni.createFrom().item(
            MultivaluedHashMap<String, String>().apply {
                add(HttpHeaders.CACHE_CONTROL, "no-cache")
                add(HttpHeaders.CONTENT_TYPE, "text/plain")
                // Make this a blocking call ;-(
                add(HttpHeaders.AUTHORIZATION, clientTokenFetcher.getClientToken().await().atMost(Duration.ofMillis(500)).authHeader)
            }
        )
    }
```

Can someone please tell me how to stay fully reactive and make it work without the blocking within the Uni?

The code above is written in kotlin, but I could also convert it into Java if that helps you...

Comment (Sgitario) [reply docs/src/main/asciidoc/rest-client-reactive.adoc:618]: The `service.getToken()` does not have to be a blocking call. This example was only intended to explain how to inject headers from a blocking service to Resteasy Rest Client. 

Also, I tried to reproduce your scenario and it worked for me:

```java
    @ApplicationScoped
    public static class Service {
        public Uni<String> getValue() {
            return Uni.createFrom().item(HEADER_VALUE);
        }
    }

    @RegisterClientHeaders(CustomReactiveClientHeadersFactory.class)
    public interface Client {
        @GET
        String getWithHeader();
    }

    public static class CustomReactiveClientHeadersFactory extends ReactiveClientHeadersFactory {

        @Inject
        Service service;

        @Override
        public Uni<MultivaluedMap<String, String>> getHeaders(MultivaluedMap<String, String> incomingHeaders) {
            return service.getValue().onItem().transform(value -> {
                MultivaluedHashMap<String, String> newHeaders = new MultivaluedHashMap<>();
                newHeaders.add(HEADER_NAME, value);
                return newHeaders;
            });
        }
    }
```

This is verified using this test https://github.com/quarkusio/quarkus/blob/main/extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/java/io/quarkus/rest/client/reactive/headers/ReactiveClientHeadersFromProviderTest.java

Therefore, it would be good to open an issue with a reproducer, so we can exactly see what is going on.

Comment (SimonScholz) [reply docs/src/main/asciidoc/rest-client-reactive.adoc:618]: Hi @Sgitario ,

thanks for your quick response:
Here you can find an example which reproduces my issue:
https://github.com/SimonScholz/ReactiveClientHeadersFactory/blob/b1a7a07c13afa9168b8897653fa27ce0f4cf3be3/src/test/kotlin/io/github/simonscholz/ReactiveResourceTest.kt#L82

In this example I'd use a reactive rest client in order to obtain a token, which is then used by the `BeverageRequestHeaderFactory` to generate a proper Authorization header.
The `ReactiveResourceTest` shows that this does not work.
I'll also create an issue later on, but wanted to inform you about my example in advance.

Thanks and regards,

Simon

Comment (SimonScholz) [reply docs/src/main/asciidoc/rest-client-reactive.adoc:618]: Hi @Sgitario,

FYI, I've now finally created an issue here: https://github.com/quarkusio/quarkus/issues/24153

Please let me know, if I can do more or if you have any questions.
